File: caves-of-zircon.java/src/main/java/rltut/FungusAi.java
package rltut;

public class FungusAi extends CreatureAi {
	private StuffFactory factory;
	private int spreadcount;
	
	public FungusAi(Creature creature, StuffFactory factory) {
		super(creature);
		this.factory = factory;
	}

	public void onUpdate(){
		if (spreadcount < 5 && Math.random() < 0.01)
			spread();
	}
	
	private void spread(){
		int x = creature.x + (int)(Math.random() * 11) - 5;
		int y = creature.y + (int)(Math.random() * 11) - 5;
		
		if (!creature.canEnter(x, y, creature.z))
			return;
		
		creature.doAction("spawn a child");
		
		Creature child = factory.newFungus(creature.z);
		child.x = x;
		child.y = y;
		child.z = creature.z;
		spreadcount++;
	}
}

File: caves-of-zircon.kotlin/src/main/kotlin/org/hexworks/cavesofzircon/ai/FungusAI.kt
package org.hexworks.cavesofzircon.ai

import org.hexworks.cavesofzircon.blocks.Creature
import org.hexworks.cavesofzircon.blocks.GameBlock
import org.hexworks.cavesofzircon.factory.EntityFactory
import org.hexworks.zircon.api.Positions
import org.hexworks.zircon.api.data.Position


class FungusAI(private val fungus: Creature,
               private val entityFactory: EntityFactory) : CreatureAI {

    private var spreadcount: Int = 0

    override fun onEnter(position: Position, block: GameBlock): Boolean {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun onUpdate() {
        if (spreadcount < 5 && Math.random() < 0.02) {
            spread()
        }
    }

    private fun spread() {
        val x = fungus.position.x + (Math.random() * 11).toInt() - 5
        val y = fungus.position.y + (Math.random() * 11).toInt() - 5
        val newPos = Positions.create(x, y)

        if (fungus.canEnter(newPos)) {
            val child = entityFactory.newFungus()
//            child.moveTo(newPos)
            spreadcount++
        }
    }
}
